extensions/amp-form/0.1/amp-form.js

```diff
@@ -1448,8 +1455,18 @@ export class AmpForm {
  * @return {boolean} Whether the form is currently valid or not.
  */
 function checkUserValidityOnSubmission(form) {
-  const elements = form.querySelectorAll('input,select,textarea,fieldset');
-  iterateCursor(elements, (element) => checkUserValidity(element));
+  const {elements} = form;
+  const formElementTagNames = ['INPUT', 'SELECT', 'TEXTAREA', 'FIELDSET'];
+  const elementsToBeChecked = Array.from(elements).filter((ele) => {
+    const tagName = ele.tagName.toUpperCase();
+    // Only allow allow-listed elements and elements must be direct descendant
+    // of form or an <input> 
+    return formElementTagNames.indexOf(tagName) > -1 && (
+      tagName === 'INPUT' ||
+      form.contains(ele)
+    );
+  });
+  iterateCursor(elementsToBeChecked, (element) => checkUserValidity(element));
```

Comment:
This would be simpler if we try to align with the signature for `querySelector` so we can replace other uses wholesale.

```js
function querySelectorAllFormElements(form, selector) {}
```

It selects those inside `form`, and `form.elements` filtered with [`matches()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/matches).